editor/src/clj/editor/dialogs.clj

```diff
@@ -660,19 +670,52 @@
          filter-term (or (:filter options)
                          (some-> filter-atom deref)
                          "")
-         initial-filtered-items (vec (filter-fn filter-term items))
+         filter-future-atom (atom nil)
+         state-atom (atom {:filter-term ""
+                           :filter-in-progress false
+                           :filtered-items []
+                           :selected-indices []})
+
+         set-filter-term
+         (fn set-filter-term [state filter-term]
+           (if (= (:filter-term state) filter-term)
+             state
+             (do (swap! filter-future-atom
+                        (fn [pending-filter-future]
+                          (thread-util/cancel-future! pending-filter-future)
+                          (future
+                            (try
+                              (let [filtered-items (vec (filter-fn filter-term items))
+                                    selected-indices (if (coll/empty? filtered-items) [] [0])]
+                                (thread-util/throw-if-interrupted!)
+                                (swap! state-atom
+                                       (fn [state]
+                                         (thread-util/throw-if-interrupted!)
+                                         (assoc state
+                                           :filter-in-progress false
+                                           :filtered-items filtered-items
+                                           :selected-indices selected-indices))))
+                              (catch InterruptedException _
+                                nil)
+                              (catch Throwable error
+                                (error-reporting/report-exception! error))))))
+                 (assoc state
+                   :filter-term filter-term
+                   :filter-in-progress true))))
```

Comment:
Typing into the filter-term field will call this function for each keypress. The filtering is performed in a `future` running in the thread pool. We keep the `future` in an `atom` so we can cancel and replace it when the term changes. Once it completes, it will update the `:filtered-items` and so on in the `state-atom` observed by the view. It is also here that we catch the `InterruptedException` thrown elsewhere if we detect that the thread has been interrupted, which will exit the `future` cleanly without doing anything.